File: avocado.R
# 4
library(readr)
# Importing the dataset
avocado <- read.csv('C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Multiple Linear Regression\\Avacado_Price.csv')
avocado <-avocado[,c(1:6)]
View(avocado)
str(avocado)
attach(avocado)
# Performing EDA
# Normal distribution
qqnorm(AveragePrice)
qqline(AveragePrice)
summary(avocado)
# Scatter plot
plot(Total_Volume, AveragePrice) # Plot relation ships between each X with Y
plot(tot_ava1, AveragePrice)
plot(tot_ava2,AveragePrice)
plot(tot_ava3, AveragePrice)
plot(Total_Bags,AveragePrice)
# Or make a combined plot
pairs(avocado)   # Scatter plot for all pairs of variables
plot(avocado)
cor(Total_Volume, AveragePrice)
cor(tot_ava1, AveragePrice)
cor(tot_ava2,AveragePrice)
cor(tot_ava3, AveragePrice)
cor(Total_Bags,AveragePrice)
cor(avocado)
# The Linear Model of interest
model.avocado <- lm(AveragePrice ~Total_Volume+tot_ava1+tot_ava2+tot_ava3+Total_Bags , data = avocado) # lm(Y ~ X)
summary(model.avocado)
model.avocadoT <- lm( AveragePrice~ Total_Volume)
summary(model.avocadoT)
model.avocado1 <- lm(AveragePrice ~ tot_ava1)
summary(model.avocado1)
model.avocado2 <-lm(AveragePrice~tot_ava2)
summary(model.avocado2)
model.avocadoc3 <- lm( AveragePrice~ tot_ava3)
summary(model.avocadoc3)
model.avocado4 <- lm(AveragePrice ~ Total_Bags)
summary(model.avocado4)
model.avocadoALL <- lm(AveragePrice ~ Total_Volume+tot_ava1+tot_ava2+tot_ava3+Total_Bags )
summary(model.avocadoALL)
# Scatter plot matrix with Correlations inserted in graph
library(GGally)
ggpairs(avocado)
# Partial Correlation matrix
library(corpcor)
library(car)
cor(avocado)
cor2pcor(cor(avocado))
plot(model.avocado)# Residual Plots, QQ-Plot, Std. Residuals vs Fitted, Cook's distance
qqPlot(model.avocado, id.n = 5) # QQ plots of studentized residuals, helps identify outliers
# Deletion Diagnostics for identifying influential obseravations
influenceIndexPlot(model.avocado, id.n = 3) # Index Plots of the influence measures
influencePlot(model.avocado, id.n = 3) # A user friendly representation of the above
# Regression after deleting the 77th observation
model.avocado1 <- lm(AveragePrice ~ Total_Volume+tot_ava1+tot_ava2+tot_ava3+Total_Bags, data = avocado[-15561, ])
summary(model.avocado1)
# Variance Inflation Factors
vif(model.avocado)  # VIF is > 10 => collinearity
# Added Variable Plots 
avPlots(model.avocado, id.n = 2, id.cex = 0.8, col = "red")
# Linear Model without Total_Volume
model.final <- lm(AveragePrice ~tot_ava1+tot_ava2+tot_ava3+Total_Bags , data = avocado)
summary(model.final)
# Linear model without Total bags and influential observation
model.final1 <- lm(AveragePrice ~Total_Volume+tot_ava1+tot_ava2+tot_ava3 , data = avocado[-810, ])
summary(model.final1)
# Added Variable Plots
avPlots(model.final1, id.n = 2, id.cex = 0.8, col = "red")
# Variance Influence Plot
vif(model.final1)
# Evaluation Model Assumptions
plot(model.final1)
plot(model.final1$fitted.values, model.final1$residuals)
qqnorm(model.final1$residuals)
qqline(model.final1$residuals)
# Subset selection
library(leaps)
lm_best <- regsubsets(AveragePrice ~ ., data = avocado, nvmax = 15)
summary(lm_best)
summary(lm_best)$adjr2
which.max(summary(lm_best)$adjr2)
coef(lm_best, 5)
lm_forward <- regsubsets(AveragePrice ~ ., data = avocado, nvmax = 15, method = "forward")
summary(lm_forward)
# Data Partitioning
n <- nrow(avocado)
n1 <- n * 0.7
n2 <- n - n1
train <- sample(1:n, n1)
test <- avocado[-train, ]
# Model Training
model <- lm(AveragePrice ~ Total_Volume+tot_ava1+tot_ava2+tot_ava3+Total_Bags, avocado[train, ])
summary(model)
pred <- predict(model, newdata = test)
actual <- test$AveragePrice
error <- actual - pred
test.rmse <- sqrt(mean(error**2))
test.rmse
train.rmse <- sqrt(mean(model$residuals**2))
train.rmse
# Step AIC
library(MASS)
stepAIC(model.avocado)

File: avocado.py
# 4
import pandas as pd
import numpy as np
# Importing the dataset
Avocado = pd.read_csv("C:\\Users\\WIN10\\Desktop\\LEARNING\\DS\\Multiple Linear Regression\\Avacado_Price.csv",encoding='unicode_escape')
Avocado.info()
Avocado.rename(columns={'AveragePrice':'Price','Total_Volume':'Volume','tot_ava1':'ava1','tot_ava2':'ava2','tot_ava3':'ava3'},inplace=True)
# Performing EDA
Avocado.describe()
#Graphical Representation
import matplotlib.pyplot as plt 
plt.bar(height = Avocado.Total_Bags, x = np.arange(1, 18250, 1))
plt.hist(Avocado.Total_Bags) #histogram
plt.boxplot(Avocado.Total_Bags) #boxplot
plt.bar(height = Avocado.Price, x = np.arange(1, 18250, 1))
plt.hist(Avocado.Price) #histogram
plt.boxplot(Avocado.Price) #boxplot
# Jointplot
import seaborn as sns
sns.jointplot(x=Avocado['Volume'], y=Avocado['Price'])
sns.jointplot(x=Avocado['ava1'], y=Avocado['Price'])
sns.jointplot(x=Avocado['ava2'], y=Avocado['Price'])
sns.jointplot(x=Avocado['ava3'], y=Avocado['Price'])
sns.jointplot(x=Avocado['Total_Bags'], y=Avocado['Price'])
# Countplot
plt.figure(1, figsize=(16, 10))
sns.countplot(Avocado['Volume'])
# Q-Q Plot
from scipy import stats
import pylab
stats.probplot(Avocado.Price, dist = "norm", plot = pylab)
plt.show()
# Correlation matrix 
Avocado.corr()
# preparing model considering all the variables 
import statsmodels.formula.api as smf # for regression model
ml1 = smf.ols('Price ~Volume+ava1+ava2+ava3+Total_Bags', data = Avocado).fit() # regression model
# Summary
ml1.summary()
# p-values for Doors are more than 0.05
# Checking whether data has any influential values 
# Influence Index Plots
import statsmodels.api as sm
sm.graphics.influence_plot(ml1)
# Studentized Residuals = Residual/standard deviation of residuals
# index 221 is showing high influence so we can exclude that entire row
# Preparing model                  
ml_new = smf.ols('Price ~ Volume+ava1+ava2+ava3+Total_Bags ', data = Avocado).fit()    
# Summary
ml_new.summary()
# Check for Colinearity to decide to remove a variable using VIF
# Assumption: VIF > 10 = colinearity
# calculating VIF's values of independent variables
#Price ~Volume+ava1+ava2+ava3+Total_Bags
rsq_vol = smf.ols('Volume ~ ava1+ava2+ava3+Total_Bags', data = Avocado).fit().rsquared  
vif_vol = 1/(1 - rsq_vol) 
rsq_1= smf.ols('ava1 ~ Volume+ava2+ava3+Total_Bags', data = Avocado).fit().rsquared  
vif_1 = 1/(1 - rsq_1)
rsq_2 = smf.ols('ava2 ~ Volume+ava1+ava3+Total_Bags', data = Avocado).fit().rsquared  
vif_2 = 1/(1 - rsq_2)
rsq_3 = smf.ols('ava3 ~ Volume+ava1+ava2+Total_Bags', data = Avocado).fit().rsquared  
vif_3 = 1/(1 - rsq_3) 
rsq_bags = smf.ols('Total_Bags~ Volume+ava1+ava2+ava3', data = Avocado).fit().rsquared  
vif_bags = 1/(1 - rsq_bags) 
# Storing vif values in a data frame
d1 = {'Variables':['Volume','ava1','ava2','ava3','Total_Bags'], 'VIF':[vif_vol, vif_1,vif_2, vif_3, vif_bags]}
Vif_frame = pd.DataFrame(d1)  
Vif_frame
# As WT is having highest VIF value, we are going to drop this from the prediction model
# Final model
final_ml = smf.ols('Price ~Volume+ava1+ava2+ava3+Total_Bags ', data = Avocado).fit()
final_ml.summary() 
# Prediction
pred = final_ml.predict(Avocado)
# Q-Q plot
res = final_ml.resid
sm.qqplot(res)
plt.show()
# Q-Q plot
stats.probplot(res, dist = "norm", plot = pylab)
plt.show()
# Residuals vs Fitted plot
sns.residplot(x = pred, y = Avocado.Price, lowess = True)
plt.xlabel('Fitted')
plt.ylabel('Residual')
plt.title('Fitted vs Residual')
plt.show()
sm.graphics.influence_plot(final_ml)
# Splitting the data into train and test data 
from sklearn.model_selection import train_test_split
Avocado_train, Avocado_test = train_test_split(Avocado, test_size = 0.2) # 20% test data
# preparing the model on train data 
model_train = smf.ols("Price ~ Volume+ava1+ava2+ava3+Total_Bags", data = Avocado_train).fit()
# prediction on test data set 
test_pred = model_train.predict(Avocado_test)
# test residual values 
test_resid = test_pred - Avocado_test.Price
# RMSE value for test data 
test_rmse = np.sqrt(np.mean(test_resid * test_resid))
test_rmse
# train_data prediction
train_pred = model_train.predict(Avocado_train)
# train residual values 
train_resid  = train_pred - Avocado_train.Price
# RMSE value for train data 
train_rmse = np.sqrt(np.mean(train_resid * train_resid))
train_rmse

